hkmc2/shared/src/main/scala/hkmc2/codegen/HandlerLowering.scala

```diff
@@ -0,0 +1,552 @@
+      Assign(res, SimpleCall(Value.Ref(resumptionSymbol), List(Value.Lit(Tree.IntLit(uid)))), rest)
+    def unapply(blk: Block) = blk match
+      case Assign(res, SimpleCall(Value.Ref(`resumptionSymbol`), List(Value.Lit(Tree.IntLit(uid)))), rest) =>
+        Some(res, uid, rest)
+      case _ => None
+  
+  object ReturnCont:
+    private val returnContSymbol = freshTmp("returnCont")
+    def apply(res: Local, uid: StateId) =
+      Assign(res, SimpleCall(Value.Ref(returnContSymbol), List(Value.Lit(Tree.IntLit(uid)))), End(""))
+    def unapply(blk: Block) = blk match
+      case Assign(res, SimpleCall(Value.Ref(`returnContSymbol`), List(Value.Lit(Tree.IntLit(uid)))), _) => 
+        Some(res, uid)
+      case _ => None
+  
+  object CallPlaceholder:
+    private val callSymbol = freshTmp("callPlaceholder")
+    def apply(res: Local, uid: StateId, canRet: Bool, c: Call, rest: Block) =
+      Assign(res, SimpleCall(Value.Ref(callSymbol), List(Value.Lit(Tree.IntLit(uid)), Value.Lit(Tree.BoolLit(canRet)))), Assign(res, c, rest))
+    def unapply(blk: Block) = blk match
+      case Assign(res, SimpleCall(Value.Ref(`callSymbol`), List(Value.Lit(Tree.IntLit(uid)), Value.Lit(Tree.BoolLit(canRet)))), Assign(_, c: Call, rest)) =>
+        Some(res, uid, canRet, c, rest)
+      case _ => None
+  
+  object StateTransition:
+    private val transitionSymbol = freshTmp("transition")
+    def apply(uid: StateId) = Return(SimpleCall(Value.Ref(transitionSymbol), List(Value.Lit(Tree.IntLit(uid)))), false)
+    def unapply(blk: Block) = blk match
+      case Return(SimpleCall(Value.Ref(`transitionSymbol`), List(Value.Lit(Tree.IntLit(uid)))), false) =>
+        S(uid)
+      case _ => N
+  
+  object FnEnd:
+    private val fnEndSymbol = freshTmp("fnEnd")
+    def apply() = Return(SimpleCall(Value.Ref(fnEndSymbol), Nil), false)
+    def unapply(blk: Block) = blk match
+      case Return(SimpleCall(Value.Ref(`fnEndSymbol`), Nil), false) => true
+      case _ => false
+  
+  private class FreshId:
+    var id: Int = 0
+    def apply() =
+      val tmp = id
+      id += 1
+      tmp
+  private val freshId = FreshId()
+  
+  // id: the id of the current state
+  // blk: the block of code within this state
+  // sym: the variable to which the resumed value should set
+  class BlockState(val id: StateId, val blk: Block, val sym: Opt[Local])
+  
+  def partitionBlock(blk: Block, labelIds: Map[Symbol, (StateId, StateId)] = Map.empty): Ls[BlockState] = 
+    // for readability :)
+    case class PartRet(head: Block, states: Ls[BlockState])
+
+    // returns (truncated input block, child block states)
+    // TODO: don't split within Match, Begin and Labels when not needed, ideally keep it intact. Need careful analysis for this
+    // blk: The block to transform
+    // labelIds: maps label IDs to the state at the start of the label and the state after the label
+    // jumpTo: what state End should jump to, if at all 
+    // freshState: uid generator
```

Comment:
Just a small suggestion to make the comment easier to parse.
```suggestion
    // * returns (truncated input block, child block states)
    // * blk: The block to transform
    // * labelIds: maps label IDs to the state at the start of the label and the state after the label
    // * jumpTo: what state End should jump to, if at all 
    // * freshState: uid generator
    // TODO: don't split within Match, Begin and Labels when not needed, ideally keep it intact. Need careful analysis for this.
```